### DISCUSSION TITLE
Trame-Paraview-Docker example returns "503 (Service Unavailable)"

### DISCUSSION THREAD

**Discussion by evdkmv** [Kitware/trame#447] [Q&A] [answered]
Hello,

I am trying to run the Trame-Paraview-Docker [example](https://github.com/Kitware/trame/tree/master/examples/07_paraview/Docker). I double-checked the contents of the files (apps.yml, requirements.txt, SimpleCone.py, Dockerfile), and they are exactly as in the example.

Then, I used the commands `docker build -t trame-paraview-docker . `  and `docker run -it --rm -p 3000:80 trame-paraview-docker`.

Unfortunately, when I open localhost:3000, I only see the 'Loading' spinner and receive the following error in the console:

```
Failed to load resource: the server responded with a status of 503 (Service Unavailable)

about to replace template try/catch on connect()

Uncaught (in promise) TypeError: Cannot read properties of null (reading 'set')
    at n (index-3684998b.js:38:1429)
    at Da (index-3684998b.js:38:1912)
```

as well as the error in the terminal (after page refresh)

```
Error handling request
Traceback (most recent call last):
  File "/deploy/server/venv/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 452, in _handle_request
    resp = await request_handler(request)
  File "/deploy/server/venv/lib/python3.10/site-packages/aiohttp/web_app.py", line 543, in _handle
    resp = await handler(request)
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/backends/aiohttp/launcher.py", line 97, in handle_post
    return await self._waitForReady(session, request)
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/backends/aiohttp/launcher.py", line 109, in _waitForReady
    if self.process_manager.isReady(session, count):
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/launcher.py", line 480, in isReady
    if not self.isRunning(id) and count < 60:
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/launcher.py", line 468, in isRunning
    return self.processes[id].poll() is None
KeyError: '42f4b4c8-cdc9-11ee-802e-0242ac110002'
```

Does anyone have an idea how to fix it?

**Comment by jourdain** [comment]
My guess is that we are missing an `ENV TRAME_CLIENT_TYPE=vue2` in the docker file on line 9.

Can you double check and submit a PR if that fix your issue?

**Reply by evdkmv** [reply]
Hello Sebastien,

Thank you for the suggestion. I added the `ENV TRAME_CLIENT_TYPE=vue2` to the Dockerfile. Unfortunately, it did not fix the issue. In the console I see: 

```
Failed to load resource: the server responded with a status of 503 (Service Unavailable). paraview/:1

app.ec4d76f4.js:1 WsError Session did not start before timeout expired. Check session logs. app.ec4d76f4.js:1
```
and the same error in the terminal as before. 

I tried to change the imports to vue3, but it was unsuccessful as well.

**Reply by evdkmv** [reply]
The similar happens with the [SingleFile](https://github.com/Kitware/trame/tree/master/examples/deploy/docker/SingleFile) and [Cone](https://github.com/Kitware/trame-app-cone) examples. With the original code`vtk_widgets.VtkGeometryRepresentation()` it works as expected. However if I substitute the code in the  app.py with e.g. this [example](https://github.com/Kitware/trame/blob/master/examples/06_vtk/01_SimpleCone/LocalRendering.py) I get the same error 

```
Failed to load resource: the server responded with a status of 503 (Service Unavailable)
app.ec4d76f4.js:1 WsError Session did not start before timeout expired. Check session logs.
```

Moreover, despite it is written `state.trame__title = "VTK Remote View - Local Rendering"`  in the new code, when I open localhost:3000 I see "trame is build by Kitware" as a title (see screenshot). 
![Screenshot 2024-02-18 at 16 08 42](https://github.com/Kitware/trame/assets/82528363/f89ef6f7-5b56-4b78-a02e-e293232b99d4)


When I try to run this [example](https://kitware.github.io/trame/examples/core/docker.html) from the docs, I get an error 


```
Starting server...
Starting the wslink launcher at
/deploy/server/venv/bin/python: Error while finding module specification for 'wslink.launcher' (ModuleNotFoundError: No module named 'wslink')
```


It feels like I am missing something.

**Reply by jourdain** [reply]
The way to figure out what's wrong would be to look at the log file in `/deploy/server/logs/launcher/*` within the image.

**Reply by evdkmv** [reply]
For a [SingleFile ](https://github.com/Kitware/trame/tree/master/examples/deploy/docker/SingleFile) example (with this [code](https://github.com/Kitware/trame/blob/master/examples/06_vtk/01_SimpleCone/RemoteRendering.py) inside app.py) in the logs I see:

```
Traceback (most recent call last):
  File "/deploy/app.py", line 13, in <module>
    from vtkmodules.vtkFiltersSources import vtkConeSource
ModuleNotFoundError: No module named 'vtkmodules'
```
I had the same error while trying the MultiFilter example. It was solved by installing vtk. 

However, according to this [discussion ](https://github.com/Kitware/trame/discussions/433) we need to use vtk-osmesa instead of vtk. When I try to run the mentioned [example](https://github.com/Kitware/trame/tree/master/examples/deploy/docker/VtkRendering), I get:


```
Starting server...
Starting the wslink launcher at
/deploy/server/venv/bin/python: Error while finding module specification for 'wslink.launcher' (ModuleNotFoundError: No module named 'wslink')
```



For [trame-paraview](https://github.com/Kitware/trame/tree/master/examples/07_paraview/Docker) example I get in the logs: 

`qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory`

Apparently Mac causing the problem. But I am pretty sure that I was able to start this docker-paraview example on Mac a month ago.

**Reply by jourdain** [reply]
I didn't know you were on mac which could indeed explain some of the issue even with the x86 emulation...

Also I'm getting confused with the various information you are providing. Especially since some of those steps/requirements are mutually exclusive. Which means if you did all of that on the same environment, it is expected to not work. 
Please open 1 discussion at a time for doing one thing while properly listing all you did.

**Reply by MntiKor** [reply]
I already encountered the error while testing trame docker image on MacOS ARM64. The only way we managed to make it work was to build ParaView OSMesa using ParaView-Superbuild on a docker arm64 image (like ubuntu), and add trame in the dockerfile. Mesa does not officially support MacOS so it's the only way to run ParaView OSMesa on Mac M1.

**Reply by evdkmv** [reply]
Thanks, Sebastien. 

Just thought these thing could be connected. I will try to properly separate issues in the future.

**Comment by jourdain** [comment]
Please @MntiKor check what could be up with your ParaView/docker example

**Reply by MntiKor** [reply]
I did know that it doesn't work on MacOS. I have a version which is way heavier because it uses ubuntu arm64 image and builds ParaView with the image build, and this one works well on KEU's Mac M1, but won't on a x86_64 machine.

**Comment by evdkmv** [comment]
@MntiKor thanks for the explanation. 

I tried to run the example on my Windows machine.

Let me summarize what I did. 
1) I created a folder "docker-paraview" on my desktop. Following the [example](https://github.com/Kitware/trame/tree/master/examples/07_paraview/Docker), the content of that folder is:
```
├── setup/
│   ├── apps.yml
│   └── requirements.txt
├── Dockerfile
└── SimpleCone.py
```

2) Opened the "docker-paraview" folder in vscode and wrote in the terminal:
`docker build -t docker-paraview-image .`
`docker run -it --rm -p 3000:80 docker-paraview-image`

So the output from /deploy/server/logs/launcher/ is `ParaView is using venv: /deploy/server/venv`. The error that I had on Mac  `qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory` is not there anymore.
        
However I get the same error as before when I go to localhost:3000: 
```
Failed to load resource: the server responded with a status of 503 (Service Unavailable)
about to replace template try/catch on connect()
Uncaught (in promise) TypeError: Cannot read properties of null (reading 'set')
    at n (index-3684998b.js:38:1429)
    at Da (index-3684998b.js:38:1912)
```
Is this example supposed to work fine on Windows? Am I missing something?

**Comment by MntiKor** [comment]
@jourdain @evdkmv when using `docker exec` on the image and trying to launch `/opt/paraview/bin/pvpython simplecone.py` , and also when running ParaView OSMesa on ubuntu22.04 official docker image, it appears that ParaView has a segmentation fault. It seems related to https://gitlab.kitware.com/paraview/paraview/-/issues/22350
I don't think there's anything we can while ParaView isn't fixed...

**Comment by jourdain** [comment, marked as answer]
I guess we should update the docker to use `FROM kitware/trame:py3.9` and download 5.11 instead
`ARG PV_URL='https://www.paraview.org/files/v5.11/ParaView-5.11.0-osmesa-MPI-Linux-Python3.9-x86_64.tar.gz'`

**Reply by evdkmv** [reply]
It works now. 

The Dockerfile was modified as follows:

```
FROM --platform=linux/amd64 kitware/trame:py3.9 #(since I am on Mac)
ARG PV_URL='https://www.paraview.org/files/v5.11/ParaView-5.11.0-osmesa-MPI-Linux-Python3.9-x86_64.tar.gz'
```
I also had to change the imports to vue3 as shown [here](https://github.com/Kitware/trame/discussions/415#discussioncomment-8150434). 

I will test it later on Windows without `--platform=linux/amd64`.

Thanks a lot @jourdain and @MntiKor for your help.

**Reply by MntiKor** [reply]
--platform should also work on Windows, but won't be mandatory. It's just to avoid Docker from downloading trame arm64 image which won't be able to run ParaView. By default Docker on Mac will download arm64 images as it's easier for qemu to emulate arm64-linux to M1 than x86_64-linux
If you have issues with performance I suggest using --platform=linux/arm64 but changing the Dockerfile so instead of downloading ParaView, it builds it from Superbuild
